File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cs
public class Solution {
    public string MinRemoveToMakeValid(string s) {
        var n = s.Length;
        var stack = new Stack<int>();
        var removeIdx = new bool[n];

        for(var i=0; i<n; i++){
            if(s[i] == '(')
                stack.Push(i);
            else if(s[i] == ')' && stack.Count > 0)
                stack.Pop();
            else if(s[i] == ')')
                removeIdx[i] = true;
        }

        while(stack.Count > 0)
            removeIdx[stack.Pop()] = true;

        var sb = new StringBuilder();
        for(var i=0; i<n; i++){
            if(removeIdx[i]) continue;
            sb.Append(s[i]);
        }

        return sb.ToString();
    }
}

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
public class Solution {
    public String minRemoveToMakeValid(String s) {
        var n = s.length();
        var stack = new Stack<Integer>();
        var removeIdx = new boolean[n];

        var chars = s.toCharArray();
        for(var i=0; i<n; i++){
            if(chars[i] == '(')
                stack.push(i);
            else if(chars[i] == ')' && stack.size() > 0)
                stack.pop();
            else if(chars[i] == ')')
                removeIdx[i] = true;
        }

        while(stack.size() > 0)
            removeIdx[stack.pop()] = true;

        var sb = new StringBuilder();
        for(var i=0; i<n; i++){
            if(removeIdx[i]) continue;
            sb.append(chars[i]);
        }

        return sb.toString();
    }
}